lib/hawkbit/hawkbit.c

```diff
@@ -0,0 +1,1221 @@
+		device_get_binding(DT_CHOSEN_ZEPHYR_FLASH_CONTROLLER_LABEL),
+		fs.offset, &info);
+	if (rc) {
+		LOG_ERR("Unable to get storage page info");
+		return -EIO;
+	}
+
+	fs.sector_size = info.size;
+	fs.sector_count = 3U;
+
+	rc = nvs_init(&fs, DT_CHOSEN_ZEPHYR_FLASH_CONTROLLER_LABEL);
+	if (rc) {
+		LOG_ERR("Storage flash Init failed");
+		return -ENODEV;
+	}
+
+	rc = nvs_read(&fs, ADDRESS_ID, &init_action_id, sizeof(init_action_id));
+	LOG_DBG("ACID: current %d, update %d", init_action_id.current,
+		init_action_id.update);
+
+	image_ok = boot_is_img_confirmed();
+	LOG_INF("Image is %s confirmed OK", image_ok ? "" : " not");
+	if (!image_ok) {
+		ret = boot_write_img_confirmed();
+		if (ret < 0) {
+			LOG_ERR("Couldn't confirm this image: %d", ret);
+			return ret;
+		}
+
+		LOG_DBG("Marked image as OK");
+		ret = boot_erase_img_bank(FLASH_AREA_ID(image_1));
+		if (ret) {
+			LOG_ERR("Failed to erase second slot");
+			return ret;
+		}
+
+		if (init_action_id.update != -1) {
+			ret = hawkbit_device_acid_update(
+				HAWKBIT_ACTION_ID_CURRENT,
+				init_action_id.update);
+		}
+	}
+
+	return ret;
+}
+
+static void response_cb(struct http_response *rsp,
+			enum http_final_call final_data,
+			enum hawkbit_http_request type)
+{
+	int ret;
+	int downloaded;
+	uint8_t *body_data = NULL;
+	static size_t body_len;
+	static size_t response_buffer_size = RESPONSE_BUFFER_SIZE;
+
+	switch (type) {
+	case HAWKBIT_PROBE:
+		if (hb_context.dl.http_content_size == 0) {
+			body_data = rsp->body_start;
+			body_len = strlen(rsp->recv_buf);
+			body_len -= (rsp->body_start - rsp->recv_buf);
+			strncpy(hb_context.response_data, body_data, body_len);
+			hb_context.dl.http_content_size = rsp->content_length;
+		}
+
+		if ((body_data == NULL) && (final_data == HTTP_DATA_MORE)) {
+			body_data = rsp->recv_buf;
+			body_len = body_len + (rsp->data_len);
+			if (body_len > response_buffer_size) {
+				response_buffer_size <<= 1;
+				hb_context.response_data = realloc(
+						   hb_context.response_data,
+						   response_buffer_size);
+				}
+			strncat(hb_context.response_data, body_data,
+					rsp->data_len);
+		}
+
+		if (final_data == HTTP_DATA_FINAL) {
+			hb_context.response_data[body_len] = '\0';
+			ret = json_obj_parse(hb_context.response_data,
+					     body_len, json_ctl_res_descr,
+					     ARRAY_SIZE(json_ctl_res_descr),
+					     &hawkbit_results.base);
+			if (ret < 0) {
+				LOG_ERR("JSON parse error");
+			}
+			body_len = 0;
+		}
+
+		break;
+
+	case HAWKBIT_CLOSE:
+	case HAWKBIT_REPORT:
+	case HAWKBIT_CONFIG_DEVICE:
+		if (strcmp(rsp->http_status, "OK") < 0) {
+			LOG_ERR("Failed to cancel the update");
+		}
+		break;
+
+	case HAWKBIT_PROBE_DEPLOYMENT_BASE:
+		if (hb_context.dl.http_content_size == 0) {
+			body_data = rsp->body_start;
+			body_len = strlen(rsp->recv_buf);
+			body_len -= (rsp->body_start - rsp->recv_buf);
+			strncpy(hb_context.response_data, body_data, body_len);
+			hb_context.dl.http_content_size = rsp->content_length;
```

Comment:
Could there now be a mismatch between respose_data length and content_length if the data was not copied fully? What would happen later in the code if that is true.